src/bindings/gpio.rs

```diff
@@ -1,64 +1,133 @@
-use rppal::gpio::Error;
-use rppal::gpio::Gpio;
-use rppal::gpio::OutputPin;
-use std::result::Result::*;
+pub mod gpio_controller;
+use gpio_controller::GpioController;
+use std::sync::atomic::{AtomicBool, Ordering};
+
 /// The enum type representing the avaiable Gpio
-/// .
-/// Each enum value is a Gpio that we can use in our project
+///
+/// Each enum value is a Gpio that you can use
+///
+// If you change the enum variant make sure to assign a integer x that respect 0 <= x < [mem::GPIO_SUPPORTED_NBRS].
+#[derive(Copy, Clone)]
 pub enum GpioPinAvailable {
-    Gpio0,
-    Gpio1,
+    Gpio0 = 0,
+    Gpio1 = 1,
 }
 
 impl GpioPinAvailable {
     /// Return the BCM GPIO pin number of [GpioPinAvailable].
-    fn to_bcm_gpio_pin_number(&self) -> u8 {
-        match *self {
-            //TODO map to the right bcm that we use
-            GpioPinAvailable::Gpio0 => 0,
-            GpioPinAvailable::Gpio1 => 1,
-        }
+    fn to_bcm_gpio_pin_number(&self) -> usize {
+        (*self) as usize
     }
 }
 
+/// Each instance represent a GPIO pin of the computer
+///
+/// ### Examples
+///
+/// let gpio_pin = GpioPin::new(& GpioPinAvailable::Gpio0)
+///
 pub struct GpioPin {
-    gpio_pin: OutputPin,
+    bmc_gpio_pin_number: usize,
 }
-
+const GPIO_PINS_IS_TAKEN: [AtomicBool; gpio_controller::GPIO_SUPPORTED_NBRS] = [
```

Comment:
There is maybe a cleaner way to initialize this array

Reply:
They might be but I already search for quite a time and didn't found it. The solution found here https://users.rust-lang.org/t/how-to-define-array-of-atomicbool/32543/3 does not feel good for me. And since it is only 32 it is not THAT big of a deal.